### PR TITLE
upstream: Null-deref on TCP health checker if setsockopt fails 

### PR DESCRIPTION
**Pull request by andrewjjenkins**
Description: Fixes a nullptr dereference that can happen if a TCP Health Checker connection gets a LocalClose.  This can happen if the connection's attempt to set socket options like TCP_KEEPIDLE fails.
Risk Level: Low.  client_ is nulled and held temporarily but client_ could already be null before.
Testing: Additional unit test added.  As well, have tested the ossfuzz repro and some other paths in the details.
Docs Changes: None
Release Notes: N/A (I'm new here - do we typically relnote fuzz bugs?)

Fixes no-longer-embargoed ossfuzz [11100](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=11100&q=envoy&colspec=ID%20Type%20Component%20Status%20Proj%20Reported%20Owner%20Summary)

This is a re-attempt at the closed #6422 .

If a TcpHealthChecker client is closed but `timeout_timer_` not disabled, then when `timeout_timer_` later fires, `client_` is dereferenced but it is now nullptr, causing a crash.

The particular ossfuzz test case is one path:

- Have a TCP health checker with a config like: `upstream_connection_options.tcp_keepalive.keepalive_time: {}`
- This causes Envoy to ask to set `TCP_KEEPIDLE` to 0, which gets `EINVAL`

```
setsockopt(53, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
setsockopt(53, SOL_TCP, TCP_KEEPIDLE, [0], 4) = -1 EINVAL (Invalid argument)
```

- A `LocalClose` is triggered but `timeout_timer_` isn't disabled.
- Later, `timeout_timer_` fires and the `client_` nullptr is dereferenced.

There is a possibility of other paths.  Some other paths like file descriptor exhaustion cause a different, intentional Envoy crash.  Limiting `/proc/sys/net/ip_local_port_range` so that `time-wait` sockets exhaust ephemeral port range causes `connect()` to fail, which triggers `RemoteClose` that is correctly handled as a failure.  So the general case I'm solving here "any call to setsockopt() fails causing a LocalClose".

### CODE REVIEW

File: source/common/upstream/health_checker_impl.cc [hunk 1]
```diff
@@ -358,21 +360,23 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onData(Buffer::Instance&
...
+      parent_.dispatcher_.deferredDelete(std::move(clientToDestroy));
     }
   } else {
     host_->setActiveHealthFailureType(Host::ActiveHealthFailureType::UNHEALTHY);
   }
 }
 
 void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onEvent(Network::ConnectionEvent event) {
-  if (event == Network::ConnectionEvent::RemoteClose) {
+  // If !client_, then we are already handling a failure/teardown
```
Comment (andrewjjenkins) [inline source/common/upstream/health_checker_impl.cc]: Key is this.  I'm using `!client_` as a signal that this is an internally-generated `RemoteClose` or `LocalClose`.  I need some way to distinguish so I know that I need to call `handleFailure()`.

(Other ways to distinguish internally-generated from externally-generated would work too)

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: The other way to do this would be to do what the other health checkers are doing which is effectively the `expect_reset` boolean. I'm fine either way.

File: source/common/upstream/health_checker_impl.cc [hunk 2]
```diff
@@ -347,7 +347,9 @@ TcpHealthCheckerImpl::TcpHealthCheckerImpl(const Cluster& cluster,
 
 TcpHealthCheckerImpl::TcpActiveHealthCheckSession::~TcpActiveHealthCheckSession() {
   if (client_) {
-    client_->close(Network::ConnectionCloseType::NoFlush);
+    Network::ClientConnectionPtr clientToDestroy(std::move(client_));
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: nit: `client_to_destroy`. Can you potentially have this in a common function with a comment on why we are doing this and then use that function in the other places we do the same thing?

File: source/common/upstream/health_checker_impl.cc (hunk 1 above)
Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: I updated to follow the `expect_reset_` boolean pattern for consistency with the other health checkers, thanks for the suggestion.

File: source/common/upstream/health_checker_impl.cc (hunk 2 above)
Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: Good idea, but following the `expect_reset_` pattern in your other comment eliminated this repetition so I think this suggestion no longer applies.

File: source/common/upstream/health_checker_impl.cc [hunk 3]
```diff
@@ -418,6 +420,7 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onInterval() {
 
     client_->connect();
     client_->noDelay(true);
+    expect_reset_ = false;
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: I think I would move this before the connect() call in case there is an inline reset? I'm not sure if that can happen or not but might as well?

File: source/common/upstream/health_checker_impl.cc [hunk 4]
```diff
@@ -431,6 +434,7 @@ void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onInterval() {
 }
 
 void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onTimeout() {
+  expect_reset_ = true;
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: I don't think this is right? Don't we want the timeout to result in a failure? Can you add a test that covers this?

File: source/common/upstream/health_checker_impl.cc (hunk 3 above)
Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: Sounds good, fixed.

File: source/common/upstream/health_checker_impl.cc (hunk 4 above)
Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: I *think* this is right, at least I'm pretty sure it preserves behavior.

I added a unit test, `DoubleTimeout` that verifies if there are two timeouts we mark the host as unhealthy; the old `Timeout` test did Timeout-then-Remote-Close which might have skipped some logic.

Here's what I think happened before I showed up in the timeout case:

1. onTimeout() fires.
```
void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onTimeout() {
  host_->setActiveHealthFailureType(Host::ActiveHealthFailureType::TIMEOUT);
  client_->close(Network::ConnectionCloseType::NoFlush);
}
```

2. `setActiveHealthFailureType` takes care of handling that we had an active health failure.

3. The `client_->close()` triggers `onEvent(Network::ConnectionEvent::LocalClose)`.

4. The old code skips calling `handleFailure` (since it's a `LocalClose`), but does `std::move(client_)` to the `deferredDelete` list.

```
void TcpHealthCheckerImpl::TcpActiveHealthCheckSession::onEvent(Network::ConnectionEvent event) {
  if (event == Network::ConnectionEvent::RemoteClose) {
    handleFailure(envoy::data::core::v2alpha::HealthCheckFailureType::NETWORK);
  }

  if (event == Network::ConnectionEvent::RemoteClose ||
      event == Network::ConnectionEvent::LocalClose) {
    parent_.dispatcher_.deferredDelete(std::move(client_));
  }

  if (event == Network::ConnectionEvent::Connected && parent_.receive_bytes_.empty()) {
    // In this case we are just testing that we can connect, so immediately succeed. Also, since
    // we are just doing a connection test, close the connection.
    // NOTE(mattklein123): I've seen cases where the kernel will report a successful connection, and
    // then proceed to fail subsequent calls (so the connection did not actually succeed). I'm not
    // sure what situations cause this. If this turns into a problem, we may need to introduce a
    // timer and see if the connection stays alive for some period of time while waiting to read.
    // (Though we may never get a FIN and won't know until if/when we try to write). In short, this
    // may need to get more complicated but we can start here.
    // TODO(mattklein123): If we had a way on the connection interface to do an immediate read (vs.
    // evented), that would be a good check to run here to make sure it returns the equivalent of
    // EAGAIN. Need to think through how that would look from an interface perspective.
    // TODO(mattklein123): In the case that a user configured bytes to write, they will not be
    // be written, since we currently have no way to know if the bytes actually get written via
    // the connection interface. We might want to figure out how to handle this better later.
    client_->close(Network::ConnectionCloseType::NoFlush);
    handleSuccess(false);
  }
}
```

5. Sometime later, `onInterval()` fires and we re-start health checking (confirmed, I believe, by unit tests that check that `enableTimer()` was called on `interval_timer_`, including the new one `DoubleTimeout`).

I could definitely be missing something or misreading here.  Or I could be preserving incorrect behavior.

File: test/common/upstream/health_checker_impl_test.cc [hunk 5]
```diff
@@ -2396,6 +2396,66 @@ TEST_F(TcpHealthCheckerImplTest, Timeout) {
...
+  connection_->raiseEvent(Network::ConnectionEvent::Connected);
+
+  EXPECT_CALL(*connection_, close(_));
+  EXPECT_CALL(*event_logger_, logEjectUnhealthy(_, _, _));
+  EXPECT_CALL(*timeout_timer_, disableTimer());
+  EXPECT_CALL(*interval_timer_, enableTimer(_));
+  timeout_timer_->callback_();
+  EXPECT_EQ(cluster_->prioritySet().getMockHostSet(0)->hosts_[0]->getActiveHealthFailureType(),
+            Host::ActiveHealthFailureType::TIMEOUT);
+  EXPECT_EQ(Host::Health::Unhealthy, cluster_->prioritySet().getMockHostSet(0)->hosts_[0]->health());
```
Comment (andrewjjenkins) [inline test/common/upstream/health_checker_impl_test.cc]: Here's where I check that two timeouts in a row cause us to mark this host unhealthy

File: test/common/upstream/health_checker_impl_test.cc [hunk 6]
```diff
@@ -2396,6 +2396,66 @@ TEST_F(TcpHealthCheckerImplTest, Timeout) {
...
+  connection_->raiseEvent(Network::ConnectionEvent::Connected);
+
+  Buffer::OwnedImpl response;
+  add_uint8(response, 1);
+  read_filter_->onData(response, false);
+
+  EXPECT_CALL(*connection_, close(_));
+  EXPECT_CALL(*event_logger_, logUnhealthy(_, _, _, true));
+  EXPECT_CALL(*timeout_timer_, disableTimer());
+  EXPECT_CALL(*interval_timer_, enableTimer(_));
```
Comment (andrewjjenkins) [inline test/common/upstream/health_checker_impl_test.cc]: Here's where I check that we've still got an `interval_timer_` hanging out, justifying the call to `interval_timer_->callback()` to trigger the second probe.

File: source/common/upstream/health_checker_impl.cc (hunk 4 above)
Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: I think this was a previous bug? Pretty sure a timeout should cause a failure to be accounted for?

Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: I'm not sure, but I can try to do `expect_reset_ = false`.  However, the Http and Grpc checkers do it the same way I have here (`expect_reset_ = true`)

https://github.com/envoyproxy/envoy/blob/4aec400e9c61c7ee03ea5e6cb5b045b6a7b5abe9/source/common/upstream/health_checker_impl.cc#L285-L296

https://github.com/envoyproxy/envoy/blob/4aec400e9c61c7ee03ea5e6cb5b045b6a7b5abe9/source/common/upstream/health_checker_impl.cc#L671-L676

and it looks like this change will break some existing unit tests as well.  I guess I thought the `setActiveHealthFailureType()` covered what we want (marking the node unhealthy after two failures).

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: I need to look at the code, something seems off here. Thanks for the detailed notes. I will get back to you in a bit.

Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: No problem, and if you decide now is the time to address bigger things, I'm happy to do what I can to help, just didn't want to go wayyy off in the weeds :-)

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: I just looked at the currently code, and unless I'm missing something I think this is a bug that was pre-existing that we should fix. I'm fairly certain that if we time out we should be ultimately immediately calling `handleFailure()`. Right? I don't think the current code does that. The reason that HTTP and gRPC work OK is that in the timeout case they will reset the stream, and that will call `handleFailure()`. 

I would add a TCP test where it only takes a single failure to cause a host to become HC failed, and verify that a single timeout does that and fires callbacks. This code will set the active failure type, but it will not cause callbacks to be invoked I think.

/wait

Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: OK, I added a unit test for a single HC timeout (check out `TcpHealthCheckerImplTest.Timeout`), verifying that `logEjectUnhealthy` and `logUnhealthy` are called and that status is set to Unhealthy.  Is this what you're talking about or are there other callbacks you're concerned about?  (None jumped out at me from other unit tests).

Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: Also I'm not sure that HTTP or gRPC work differently; I guess I'd say to my eyes they're behaving very similarly to the way this PR is currently treating TCP HC timeouts.

HTTP calls `client_->close()` immediately after setting `expect_reset_ = true`:

https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L291-L294

and the HTTP `onEvent()` doesn't call `handleFailure()` on a local close either:(https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L178-L186

and looks like gRPC sets `expected_reset_ = true`: https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L671

and then uses this specifically to bypass calling `handleFailure()`:

https://github.com/envoyproxy/envoy/blob/bc4f1d7396a4634f81e7aecc2e18543afbdcc15b/source/common/upstream/health_checker_impl.cc#L597-L601

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: Yeah I think this is all broken, which is pretty surprising to me. @snowp can you PTAL at this to make sure I'm not crazy?

Comment (snowp) [reply source/common/upstream/health_checker_impl.cc]: Yeah I agree that this seems broken for HTTP/gRPC as well. We should probably be explicitly calling handleFailure or let it flow through to the reset handler

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: Thanks @snowp. @andrewjjenkins can you fix those cases also and make sure we have tests that the callbacks are called after a single timeout? Thank you so much!

Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: OK, I pushed additional unit tests for HTTP and gRPC so I think I'm confirming that we are marking unhealthy and calling callbacks after one timeout:

- HTTP: `HttpHealthCheckerImplTest.Timeout`
- gRPC: `GrpcHealthCheckerImplTest.Timeout`
- TCP: `TcpHealthCheckerImplTest.Timeout`

So just double-checking, AFAICT we are calling the callbacks in all cases. I'm trying to confirm we believe the code as-is in this PR is giving us the correct result (mark unhealthy, call callbacks, even after one timeout).

It sounds like the shortcoming y'all are seeing is related to not calling `handleFailure()`, but it looks to me like that's done by the base impl:
https://github.com/envoyproxy/envoy/blob/6178aeac7b446e3d5121bcc06ae85a6962a72ebb/source/common/upstream/health_checker_base_impl.cc#L361-L364

So I guess I'm not sure what to change - are we trying to pull the call to `handleFailure()` out of `onTimeoutBase()` or something else?

Comment (mattklein123) [reply source/common/upstream/health_checker_impl.cc]: Ah OK, sorry, that's what @snowp and I were missing from code inspection. There is no bug here because of calling handleFailure() in the base class timeout function. So I think there is nothing to fix. Let's keep the extra tests though unless they are duplicates of something else. Given that, is this PR ready to review again? Sorry for the runaround!

Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.cc]: No problem, always a pleasant surprise when things are less broken than you had guessed :-)

Yes, I think it's ready-to-review.  I am OK removing some of the single-timeout tests if we think there's not enough value to justify carrying them along, let me know what you think after looking at the code.

File: source/common/upstream/health_checker_impl.h [hunk 7]
```diff
@@ -252,6 +252,9 @@ class TcpHealthCheckerImpl : public HealthCheckerImplBase {
     TcpHealthCheckerImpl& parent_;
     Network::ClientConnectionPtr client_;
     std::shared_ptr<TcpSessionCallbacks> session_callbacks_;
+    // If true, stream reset was initiated by us, not e.g. remote reset.
+    // In this case healthcheck status already reported, only state cleanup required.
+    bool expect_reset_{};
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.h]: nit: can you make this `expect_close_` and update the comment above to talk about close vs. reset? There is no reset happening for TCP.

File: test/common/upstream/health_checker_impl_test.cc [hunk 8]
```diff
@@ -2187,9 +2234,9 @@ class TcpHealthCheckerImplTest : public testing::Test {
     }
     )EOF";
 
-    health_checker_.reset(new TcpHealthCheckerImpl(*cluster_, parseHealthCheckFromV1Json(json),
-                                                   dispatcher_, runtime_, random_,
-                                                   HealthCheckEventLoggerPtr(event_logger_)));
+    health_checker_.reset(
+        new TcpHealthCheckerImpl(*cluster_, parseHealthCheckFromV1Json(json.str()), dispatcher_,
```
Comment (mattklein123) [inline test/common/upstream/health_checker_impl_test.cc]: As long as you are in here would you mind updating these tests to use v2 YAML instead of JSON as @derekargueta has been doing elsewhere? This would really help our v1 deprecation efforts.

File: source/common/upstream/health_checker_impl.h (hunk 7 above)
Comment (andrewjjenkins) [reply source/common/upstream/health_checker_impl.h]: Done

File: test/common/upstream/health_checker_impl_test.cc (hunk 8 above)
Comment (andrewjjenkins) [reply test/common/upstream/health_checker_impl_test.cc]: Done, these were the last users of parseHealthCheckFromV1Json so I removed that too

File: test/common/upstream/health_checker_impl_test.cc [hunk 9]
```diff
@@ -2170,65 +2211,55 @@ class TcpHealthCheckerImplTest : public testing::Test {
...
+    std::ostringstream yaml;
+    yaml << R"EOF(
+    timeout: 1s
+    interval: 1s
+    unhealthy_threshold: )EOF"
+         << unhealthy_threshold << R"EOF(
+    healthy_threshold: 2
+    tcp_health_check:
+      send:
+        text: "01"
```
Comment (andrewjjenkins) [inline test/common/upstream/health_checker_impl_test.cc]: If I'm reading the proto file correctly, this is the right way to send binary encoded as hex stream (and the unit tests that use this still pass), but wouldn't mind a double-check here.

https://github.com/envoyproxy/envoy/blob/31485b5d007a87352853ca16d5ee5b91e400d85b/api/envoy/api/v2/core/health_check.proto#L80-L81
